ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -938,6 +949,339 @@ private void updateClusterMapInHelix(boolean startValidatingClusterManager) thro
+    if (!dryRun) {
+      dcAdmin.addResource(clusterName, resourceName, idealState);
+      info("[{}] Added {} new partitions under resource {} in datacenter {}", dcName.toUpperCase(),
+          partitionsUnderResource.size(), resourceName, dcName);
+    } else {
+      info("[{}] Under DryRun mode, {} new partitions are added to resource {} in datacenter {}", dcName.toUpperCase(),
+          partitionsUnderResource.size(), resourceName, dcName);
+    }
+    resourcesAdded.getAndIncrement();
+  }
+
+  /**
+   * Build a new IdealState for the given resource name and partition layout
+   * @param dcName The name of the datacenter
+   * @param resourceName The name of the newly created resource
+   * @param partitionsUnderResource The partition layout
+   * @return The new IdealState
+   */
+  private IdealState buildIdealState(String dcName, String resourceName,
+      List<Map.Entry<String, Set<String>>> partitionsUnderResource) {
+    IdealState idealState = new IdealState(resourceName);
+    idealState.setStateModelDefRef(stateModelDef);
+    info("[{}] Adding partitions for next resource {} in {}. {}.", dcName.toUpperCase(), resourceName, dcName,
+        dryRun ? "Actual IdealState is not changed as dry run" : "IdealState is being updated");
+    updateIdealStatePartitions(idealState, partitionsUnderResource);
+    if (!idealState.isValid()) {
+      throw new IllegalStateException("IdealState could not be validated for new resource " + resourceName);
+    }
+    return idealState;
+  }
+
+  /**
+   * Update the partition layout in the given ideal state.
+   * @param idealState The ideal state to change partition layout
+   * @param partitionsUnderResource The partition layout.
+   */
+  private void updateIdealStatePartitions(IdealState idealState,
+      List<Map.Entry<String, Set<String>>> partitionsUnderResource) {
+    for (Map.Entry<String, Set<String>> entry : partitionsUnderResource) {
+      String partitionName = entry.getKey();
+      ArrayList<String> instances = new ArrayList<>(entry.getValue());
+      Collections.shuffle(instances);
+      idealState.setPreferenceList(partitionName, instances);
+    }
+    idealState.setNumPartitions(idealState.getPartitionSet().size());
+    idealState.setReplicas(ResourceConfig.ResourceConfigConstants.ANY_LIVEINSTANCE.name());
+  }
+
+  /**
+   * When resources and static partition layout are both FULL_AUTO compatible, this method would iterate through the new
+   * partitions and find corresponding resources for them if possible.
+   * Before calling this method, we should already verify that each instance has only one assigned resource. So if a
+   * partition's instances are already assigned to the same resource, then this partition would be assigned to this resource
+   * as well. If a partition's instances are assigned to different resources, then this method would fail.
+   * @param dcName The name of the datacenter
+   * @param partitionsToInstancesInDc The map of partition layout.
```

Comment:
QQ. When we return from this method, would this remove the partitions whose instances are already part of a resource? 

Minor: Can you please help add comment to say if a parameter is used as input or both input/output (modified in the method)?

Reply:
updated